### DISCUSSION TITLE
ODBC Hook Providing Driver in Extra

### DISCUSSION THREAD

**Discussion by abajpai001** [apache/airflow#34373] [General]
Hi,

In my use case I have to open ODBC connection using OdbcHook. 
I am providing Driver path in extra in connection tab but getting below warning 
**[2023-09-14, 11:47:06 UTC] {odbc.py:117} WARNING - You have supplied 'driver' via connection extra but it will not be used. In order to use 'driver' from extra you must set airflow config setting `allow_driver_in_extra = True` in section `providers.odbc`. Alternatively you may specify driver via 'driver' parameter of the hook constructor or via 'hook_params' dictionary with key 'driver' if using SQL operators.**
now when i try adding this setting in airflow config through UI , it is not letting me add section naming providers.odbc 
**Sections must not contain hyphens ("-"), periods ("."), opening square brackets ("["), or closing square brackets ("]").**

I had another way of adding this parameter in Environment Variables using **AIRFLOW__PROVIDERS_ODBC__ALLOW_DRIVER_IN_EXTRA** parameter 
this is also giving me error 
**Environment variable names cannot override an Airflow software configuration (i.e., match the regular expression "AIRFLOW__[A-Z0-9_]+__[A-Z0-9_]+").**

**Comment by Taragolis** [comment]
> [2023-09-14, 11:47:06 UTC] {odbc.py:117} WARNING - You have supplied 'driver' via connection extra but it will not be used. In order to use 'driver' from extra you must set airflow config setting allow_driver_in_extra = True in section providers.odbc. Alternatively you may specify driver via 'driver' parameter of the hook constructor or via 'hook_params' dictionary with key 'driver' if using SQL operators.

This feature disabled by default due to [CVE-2023-34395](https://nvd.nist.gov/vuln/detail/CVE-2023-34395) 

If you need to use and want to enable globally, you need to setup in your `airflow.cfg`
```ini
[providers.odbc]
allow_driver_in_extra = True
```

Or set environment variables `AIRFLOW__PROVIDERS_ODBC__ALLOW_DRIVER_IN_EXTRA=True`

But seems like you use some managed Airflow service which doesn't allow you to do this? Am I right?

**Reply by abajpai001** [reply]
Yes we are using GCP composer environment where Airflow runs.

**Reply by Taragolis** [reply]
I would recommend ask GCP support how you could define custom settings or try to find this information in [Composer User Guide/Documentation](https://cloud.google.com/composer/docs).

As alternative you could provide Driver as part of Hook/SQLOperator parameters as it mentioned in driver, see sample from tests (the better we have for ODBC Hook):

https://github.com/apache/airflow/blob/438ba41e142593f2b0916893eccbd08fbe4d277b/tests/providers/odbc/hooks/test_odbc.py#L63-L76

**Reply by abajpai001** [reply]
Hi - As per the suggestion, we connected with GCP and figured a way to add the property in airflow configuration as given in Airflow documentation(attached image below).
But my DAG is still failing to read connection extra(2nd image attached)

![AirflowConfiguration](https://github.com/apache/airflow/assets/112093168/a747089e-9fbe-4ad5-8e6f-4f6464c8cb3c)
![AIrflowDagFailure](https://github.com/apache/airflow/assets/112093168/4963ed4a-02f6-4ca1-9329-21e0ebbf2660)
